### PR TITLE
Bug #114637 fix: Backend:SQL injection flow on countries view ->vulne…

### PR DESCRIPTION
**Pull request by deepa-g**
- [ ] Bug #114637 Fix: Backend: SQL injection flow on countries view ->vulnerable parameter is the client.

- [ ] Bug #114631 Fix: BackEnd: SQL injection on regions view on client parameter

### CODE REVIEW

File: administrator/models/cities.php [hunk 1]
```diff
@@ -152,15 +153,28 @@ protected function getListQuery()
...
+			}
+		}
 
 		$query->select('c.country');
 		$query->join('LEFT', '`#__tj_country` AS c ON c.id=a.country_id');
-		$query->where('c.' . $client .' = 1');
+
+		if (!empty($client))
+		{
+			if (in_array($client, $clientArray))
```
Comment (manojLondhe) [inline administrator/models/cities.php]: @deepa-g 
instead of this check if a db column exists for given client
do this everywhere in this MR

Comment (deepa-g) [reply administrator/models/cities.php]: Done

File: administrator/models/cities.php [hunk 2]
```diff
@@ -140,9 +140,16 @@ protected function getStoreId($id = '')
 	protected function getListQuery()
 	{
 		// Create a new query object.
+		$app = JFactory::getApplication();
 		$db = $this->getDbo();
 		$query = $db->getQuery(true);
-		$client = JFactory::getApplication()->input->get('client', '', 'STRING');
+		$client = $app->input->get('client', '', 'STRING');
+		$dbprefix = $app->get('dbprefix');
```
Comment (ahghatol) [inline administrator/models/cities.php]: @deepa-g add quote to table $query->from($db->quoteName('#_table_name'));

File: administrator/models/cities.php [hunk 3]
```diff
@@ -152,15 +159,28 @@ protected function getListQuery()
 		);
 
 		$query->from('`#__tj_city` AS a');
-		$query->select('a.' . $client .' AS state');
+
+		if (!empty($client))
+		{
+			if (in_array($client, $clientArray))
+			{
+				$query->select('a.' . $client . ' AS state');
```
Comment (ahghatol) [inline administrator/models/cities.php]: @deepa-g There is not quote to state column

File: administrator/models/countries.php [hunk 4]
```diff
@@ -105,16 +105,24 @@ protected function getStoreId($id = '')
...
 	{
 		// Create a new query object.
+		$app = JFactory::getApplication();
 		$db = $this->getDbo();
 		$query = $db->getQuery(true);
-		$client = JFactory::getApplication()->input->get('client', '', 'STRING');
+		$client = $app->input->get('client', '', 'STRING');
+		$dbprefix = $app->get('dbprefix');
+
+		$query = "SHOW COLUMNS FROM " . $dbprefix . "tj_country";
```
Comment (ahghatol) [inline administrator/models/countries.php]: @deepa-g Refer 1st comment

File: administrator/models/countries.php [hunk 5]
```diff
@@ -124,18 +132,36 @@ protected function getListQuery()
 		);
 		$query->from('`#__tj_country` AS a');
 
-		$query->select('a.' . $client .' AS state');
+		if (!empty($client))
+		{
+			if (in_array($client, $clientArray))
+			{
+				$query->select('a.' . $client . ' AS state');
```
Comment (ahghatol) [inline administrator/models/countries.php]: Refer 2nd comment

File: administrator/models/fields/countries.php [hunk 6]
```diff
@@ -42,17 +44,27 @@ class JFormFieldCountries extends JFormFieldList
 	 */
 	protected function getOptions()
 	{
+		$app = JFactory::getApplication();
 		$db = JFactory::getDbo();
 		$query = $db->getQuery(true);
```
Comment (ahghatol) [inline administrator/models/fields/countries.php]: @deepa-g Scan your entier changes as per above comments

File: administrator/models/cities.php (hunk 2 above)
Comment (deepa-g) [reply administrator/models/cities.php]: Done.

File: administrator/models/fields/countries.php (hunk 6 above)
Comment (deepa-g) [reply administrator/models/fields/countries.php]: Done

File: administrator/models/countries.php (hunk 4 above)
Comment (deepa-g) [reply administrator/models/countries.php]: Changes done.

File: administrator/models/cities.php [hunk 7]
```diff
@@ -66,6 +66,9 @@ protected function populateState($ordering = null, $direction = null)
 		// Initialise variables.
 		$app = JFactory::getApplication('administrator');
 
+		$client = $app->input->get('client', '', 'STRING');
```
Comment (coolbung) [inline administrator/models/cities.php]: Is it needed to write code in populateState at all ?

If your request variables are named correctly, and the constructor has a list of filterable fields, the state should get auto-populated.

Isn't that how it works @ahghatol 

cc @manojLondhe @thite-amol

Comment (thite-amol) [reply administrator/models/cities.php]: Joomla will automatically set the state for the variables if they are sent through the filters array. This is mostly used for the stateful list views. All other variables need to be populated in the populate state.

File: administrator/models/cities.php [hunk 8]
```diff
@@ -140,27 +143,36 @@ protected function getStoreId($id = '')
 	protected function getListQuery()
 	{
 		// Create a new query object.
+		$app = JFactory::getApplication();
```
Comment (thite-amol) [inline administrator/models/cities.php]: @deepa-g This should not in the getListQuery

File: administrator/models/cities.php [hunk 9]
```diff
@@ -140,27 +143,36 @@ protected function getStoreId($id = '')
...
 		$db = $this->getDbo();
 		$query = $db->getQuery(true);
-		$client = JFactory::getApplication()->input->get('client', '', 'STRING');
+		$client = $this->getState('client');
+		$dbprefix = $app->get('dbprefix');
```
Comment (thite-amol) [inline administrator/models/cities.php]: @deepa-g @coolbung @manojLondhe Can we have a different place for dbprefix like a class variable?

File: administrator/models/cities.php [hunk 10]
```diff
@@ -140,27 +143,36 @@ protected function getStoreId($id = '')
...
+		$query = "SHOW COLUMNS FROM " . $dbprefix . "tj_city";
```
Comment (thite-amol) [inline administrator/models/cities.php]: @deepa-g sanitize the user input.

File: administrator/models/cities.php (hunk 9 above)
Comment (manojLondhe) [reply administrator/models/cities.php]: yes, define on class level and then set it in the constructor

File: administrator/models/cities.php (hunk 7 above)
Comment (deepa-g) [reply administrator/models/cities.php]: Fixes Done.

Comment (manojLondhe) [reply administrator/models/cities.php]: this is not a filter, so this looks ok to me

Comment (coolbung) [reply administrator/models/cities.php]: The filter still needs to change to CMD. STRING is too permissive

**Pull request merged**